generator/src/main/java/org/apache/kafka/message/MessageDataGenerator.java

```diff
@@ -624,133 +829,357 @@ private String readFieldFromStruct(FieldType type, String name) {
-            }
+            buffer.printf("if (%s != %s) {%n", field.camelCaseName(), fieldDefault(field));
         }
     }
 
     private void generateClassToStruct(String className, StructSpec struct,
                                        Versions parentVersions) {
         headerGenerator.addImport(MessageGenerator.STRUCT_CLASS);
         buffer.printf("@Override%n");
-        buffer.printf("public Struct toStruct(short version) {%n");
+        buffer.printf("public Struct toStruct(short _version) {%n");
         buffer.incrementIndent();
-        if (generateInverseVersionCheck(parentVersions, struct.versions())) {
-            buffer.incrementIndent();
-            headerGenerator.addImport(MessageGenerator.UNSUPPORTED_VERSION_EXCEPTION_CLASS);
-            buffer.printf("throw new UnsupportedVersionException(\"Can't write " +
-                "version \" + version + \" of %s\");%n", className);
-            buffer.decrementIndent();
-            buffer.printf("}%n");
-        }
+        VersionConditional.forVersions(parentVersions, struct.versions()).
+            allowMembershipCheckAlwaysFalse(false).
+            ifNotMember(__ -> {
+                headerGenerator.addImport(MessageGenerator.UNSUPPORTED_VERSION_EXCEPTION_CLASS);
+                buffer.printf("throw new UnsupportedVersionException(\"Can't write " +
+                    "version \" + _version + \" of %s\");%n", className);
+            }).
+            generate(buffer);
         Versions curVersions = parentVersions.intersect(struct.versions());
-        if (curVersions.empty()) {
-            throw new RuntimeException("Version ranges " + parentVersions +
-                " and " + struct.versions() + " have no versions in common.");
-        }
-        buffer.printf("Struct struct = new Struct(SCHEMAS[version]);%n");
+        headerGenerator.addImport(MessageGenerator.TREE_MAP_CLASS);
+        buffer.printf("TreeMap<Integer, Object> _taggedFields = null;%n");
+        VersionConditional.forVersions(messageFlexibleVersions, struct.versions()).
+            ifMember(__ -> {
+                buffer.printf("_taggedFields = new TreeMap<>();%n");
+            }).
+            generate(buffer);
+        buffer.printf("Struct struct = new Struct(SCHEMAS[_version]);%n");
         for (FieldSpec field : struct.fields()) {
-            generateFieldToStruct(field, curVersions);
+            VersionConditional.forVersions(field.versions(), curVersions).
+                alwaysEmitBlockScope(field.type().isArray()).
+                ifMember(presentVersions -> {
+                    VersionConditional.forVersions(field.taggedVersions(), presentVersions).
+                        ifNotMember(presentAndUntaggedVersions -> {
+                            generateFieldToStruct(field, presentAndUntaggedVersions);
+                        }).
+                        ifMember(presentAndTaggedVersions -> {
+                            generateNonDefaultValueCheck(field);
```

Comment:
I think there's a bug in the handling of nullable arrays when the default is not null. For example, consider the following field:

```json
    { "name": "field2", "type": "[]BlahType",
      "versions": "1+", "taggedVersions": "1+", "tag": 1,
      "nullableVersions": "1+",
      "fields": [
          { "name": "wootId", "versions": "1+", "type": "int32" },
      ]
    }
```
This results in the following code:

```java
        if (_version >= 1) {
            if (!field2.isEmpty()) {
                if (field2 == null) {
                    _taggedFields.put(1, null);
                } else {
                    Struct[] _nestedObjects = new Struct[field2.size()];
                    int i = 0;
                    for (BlahType element : this.field2) {
                        _nestedObjects[i++] = element.toStruct(_version);
                    }
                    _taggedFields.put(1, _nestedObjects);
                }
            }
        }
```
The null check should come first. Seems like the default value optimization needs to take into account nullable values. The same bug affects `size`.

In general, we probably need more testing, especially for default value handling.

Reply:
Thanks for finding this.  It might be better to address it in a follow on, since the fix could get complicated.  I'll push what I have for now.